src/java.base/share/classes/jdk/crac/Resource.java

```diff
@@ -28,23 +28,50 @@
 
 /**
  * An interface for receiving checkpoint/restore notifications.
- *
- * <p>The class that is interested in receiving a checkpoint/restore notification
+ * <p>
+ * The class that is interested in receiving a checkpoint/restore notification
  * implements this interface, and the object created with that class is
  * registered with a {@code Context}, using {@code register} method.
+ * <p>
+ * All notifications before checkpoint and after restore are performed in a
+ * single thread, therefore the resource does not have to guard against
+ * concurrent invocations of these methods. However, other threads might be
+ * running during these notifications, therefore it is up to the implementation
+ * to protect against concurrent access.
  */
 public interface Resource {
 
     /**
      * Invoked by a {@code Context} as a notification about checkpoint.
+     * The resource should not depend on the state of any other resource; when
+     * this method is invoked it is possible that some other resource's
+     * notification has thrown an error and/or its
+     * {@link #afterRestore(Context)} method has been already called.
+     * <p>
+     * The order of notification is subject to the {@link Context}
+     * implementation, e.g. for the global context see
+     * {@link Core#getGlobalContext()}.
      *
      * @param context {@code Context} providing notification
      * @throws Exception if the method have failed
      */
     void beforeCheckpoint(Context<? extends Resource> context) throws Exception;
 
     /**
-     * Invoked by a {@code Context} as a notification about restore.
+     * Invoked by a {@code Context} both as a notification about restore or
+     * when the checkpoint cannot be performed (e.g. due to this or some other
+     * resource throwing an exception when {@link #beforeCheckpoint(Context)
+     * beforeCheckpoint}.
+     * Therefore, the resource should not have assumptions about it state; it
```

Comment:
Resource can be sure the beforeCheckpoint was called, and object is exactly in the state at which the beforeCheckpoint has leaved it.

Reply:
I'll reword. The meaning here was to guide user to handling unexpected states, e.g. if the beforeCheckpoint is locking a lock this should not blindly unlock.